Code Coverage Support

nicholas-maltbie opened the pull request:
#### Changes

- Added basic code coverage support to project
- Added code coverage version 1.0.1 to the example project packages
- Added three new parameters for code coverage
  - `enableCodeCoverage` Should code coverage results be collected for the test run.
  - `coverageAssemblyFilters` Optional coverage assembly filters when collecting code coverage.
  - `coverageResultsPath` Path to write coverage results to when collecting code coverage.
- Added new output for code coverage
  - `coverageResultsPath` Path where the code coverage results are stored.
- Added some tests to validate that the code coverage results are generated as expected. 

Included the following scenarios for test coverage:
1. Include flag for `enableCodeCoverage` - Will generate test results based on testMode. If EditMode or PlayMode are selected, it will generate coverage for the tests run. If All is selected, it will generate coverage from the combined EditMode and PlayMode tests from the [Generate combined report from EditMode and PlayMode tests](https://docs.unity3d.com/Packages/com.unity.testtools.codecoverage@1.0/manual/CoverageBatchmode.html#generate-combined-report-from-editmode-and-playmode-tests) together described by Issue #181
2. Allow to write results to the `coverageResultsPath` to create a html and XML summary of test reports. Default parameter value is CodeCoverage.
3. Optional parameter for `coverageAssemblyFilters` that will include a `assemblyFilters` parameter for the `-coverageOptions` flag if a value is provided. Otherwise, it will use no value and not include the `assemblyFilters` flag. 

#### Checklist

- [x] Read the contribution [guide](../CONTRIBUTING.md) and accept the [code](../CODE_OF_CONDUCT.md) of conduct
- [x] Readme (updated or not needed)
  - [x] Updated documentation of parameters for job and action.
- [x] Tests (added, updated or not needed)
  - [x] Added validation and tests for new inputs and outputs
  - [x] Added tests for new test workflow and uploading artifacts

webbertakken commented on action.yml:
```diff
@@ -20,6 +20,18 @@ inputs:
+    required: false
+    default: 'false'
+    description: 'Should code coverage results be collected for the test run.'
+  coverageAssemblyFilters:
+    required: false
+    default: ''
+    description: 'Optional coverage assembly filters when collecting code coverage.'
```
Would it be possible to come up with sensible defaults instead of exposing this in our public api in the first iteration?

nicholas-maltbie replied:
It would probably be best to include an option for `-coverageOptions` with `enableCyclomaticComplexity;generateHtmlReport;generateBadgeReport` as the default argument. And an end user could add in `assemblyFilters:+MyAssembly` if needed. Makes much more sense to include it this way.

The only reason this would become a bit difficult is that combining results from playmode and test mode can be quite difficult, but I'll keep looking into it for now.

webbertakken commented on action.yml:
```diff
@@ -20,6 +20,18 @@ inputs:
+    required: false
+    default: ''
+    description: 'Optional coverage assembly filters when collecting code coverage.'
+  coverageResultsPath:
+    required: false
+    default: 'CodeCoverage'
+    description: 'Path to write coverage results to when collecting code coverage.'
```
Would it be possible to follow convention (folder is usually called `[Cc]overage` from projects across a few languages I've seen).

Perhaps it could be stored inside the `artifacts` folder that we already create containing the test results (many people have already ignored this in their `.gitignore`)?

nicholas-maltbie replied:
Sure, adding it to the `[Cc]overage` folder (Unity's project's default folder is called `CodeCoverage`) as a default arugmnet is a great idea. The one reason I would try to avoid putting it in the `artifacts` is that it can be a fairly large build. We can make the default `coverage` but let the user configure them as different path (even `artifacts` if they want) to allow them to control each export individually.

webbertakken replied:
Ah if Unity has `CodeCoverage` as a default, then we should probably follow their convention.

> The one reason I would try to avoid putting it in the `artifacts` is that it can be a fairly large build.

`unity-builder` doesn't output to `artifacts` folder though. `unity-test-runner` currently puts all it's outputs in `artifacts` folder. It's not more than a few megs large.

If with that information you still think it's a bad idea to expose it as part of the existing output folder, then I suggest we hardcode `CodeCoverage` for the folder and `output.coverage` for the output parameter in favour of "great defaults over many configuration options", at least for now.

Let me know what you think.

nicholas-maltbie replied:
I agree with this, let's keep the default the same as unity's for now and not make it configurable (at least for this version). I'll remove the input parameter but keep the output parameter. Unity has UI for configuring the code coverage so let's use that as a guide for what we should make as all the configurable parameters, although we can leave that to a future PR.

![image](https://user-images.githubusercontent.com/3240136/164141052-800c7db9-e12c-4c3e-8d3a-5ff43d9478f7.png)

They have the configurable parameters:
* `Results Location`: output location which we can make configurable later if necessary... It defaults to `CodeCoverage` so we can leave that as the output path for now and add it as an input parameter if we need. 
* `History Location`: which is where the historical results of tests are saved, not as useful for reporting so we can probably leave it out for now.
* `Enable Code Coverage`: which we can probably always leave enabled as it doesn't hurt anything in the test build. If an end user wants to run a perf test, they should probably make a compiled build first anyway so not harm in including the debug flag. 
* `Included Assemblies`: This is the assembly filter argument; this could probably be made this something a user could edit by the existing `coverageParameters` argument.
* `Included Paths`/`Excluded Paths`: can also be configured from the `coverageParameters` argument. Slightly complex for an end user but it will default to all the code included in the `Assets` folder so unless they included custom libraries, they should be fine.
* `Generate HTML Report`: This is very useful for visually seeing the coverage but is quite a large file set if the user has a large project or lots of imported libraries. So, this might be a flag we want to add in the future. Although this can also be configured from the from the `coverageParameters` argument as well so maybe just leave it for more advanced users if they want. 
* `Generate Summary Badges`: This generates the badges for showing whether it passed or not as an icon, fairly lightweight so we can leave it always enabled.
* `Generate History`: Again, history is useful for debugging but not as useful in reporting on a single build so we can leave it disabled. 
* `Generate Additional Metrics`: I think this is something most users wouldn't need but could be added with the `coverageParameters` argument if the user wants to.


Sorry for the long explanation, but basically I suggest we add these parameters now:
* `coverageParametrers` for the custom coverage parameters via the `-coverageOptions` that allows us to configure most all of the code coverage settings. 
* Remove the `coverageResultsPath` as a default of `[Project]/CodeCoverage` should be sufficient for most users and aligns with the unity default. 
* Keep the output path of results as you suggested in case we add configuration in the future!

With just adding one parameter for `coverageParametrers` (which I will rename to `coverageOptions` to align with the flag name), the user can configure almost everything that they would want to! The parameter would be a bit difficult to configure sometimes but most users probably want the HTML report or could ignore it in their file upload if they don't want to save it.

webbertakken commented on dist/steps/run_tests.sh:
```diff
@@ -169,6 +205,44 @@ if [ "$PLAY_MODE" = "true" ]; then
 fi
 
+#
+# Combine test results if needed
+#
+COMBINE_EXIT_CODE=0
+if [ "$EDIT_MODE" = "true" ] && [ "$PLAY_MODE" = "true" ] && [ "$ENABLE_CODE_COVERAGE" = "true" ]; then
+  echo ""
+  echo "##############################"
+  echo "# Combining Coverage Results #"
+  echo "##############################"
+  echo ""
+  unity-editor \
+    -batchmode \
+    -debugCodeOptimization \
+    -enableCodeCoverage \
+    -logFile "$FULL_ARTIFACTS_PATH/combine_coverage.log" \
+    -projectPath "$UNITY_PROJECT_PATH" \
+    -coverageResultsPath "$FULL_COVERAGE_RESULTS_PATH" \
+    -coverageOptions "generateHtmlReport;generateBadgeReport" \
+    -quit
+
+  # Catch exit code
+  COMBINE_EXIT_CODE=$?
+
+  # Print unity log output
+  cat "$FULL_ARTIFACTS_PATH/combine_coverage.log"
+
+  # Display results
+  if [ $COMBINE_EXIT_CODE -eq 0 ]; then
+    echo "Combine coverage succeeded, no failures occurred";
+  elif [ $COMBINE_EXIT_CODE -eq 2 ]; then
+    echo "Combine coverage, some tests failed";
+  elif [ $COMBINE_EXIT_CODE -eq 3 ]; then
+    echo "Combine coverage (other failure)";
+  else
+    echo "Unexpected exit code $COMBINE_EXIT_CODE";
+  fi
+fi
+
```
I'd like to think a bit more before committing to having another duplicated block in bash.

Ideally we would not need 3 separate blocks of code just to run the editor for playMode, editMode, or both.

nicholas-maltbie replied:
Given the previous comment, I might adjust this so that we have run commands. And if the player selected "All" as the test platform, then in the javascript code, we could generate the three blocks of parameters to run (for edit mode, playmode, and then the third one to combine them).

nicholas-maltbie replied:
Ok, I wrote some adjustments to the `run_tests.sh` file. We could determine most of those parameters from the javascript however, but we need to change the logic to run the unity program a variable number of times. I achieved this by making the `TEST_PLATFORMS` parameter which will have a list of test platforms separated by a `;` with a special platform `COMBINE_RESULTS` for combining the results of previous tests. 

If we really don't want to make any changes to the bash script, then this change can be avoided. Just an idea of how we could implement this differently.

webbertakken replied:
Long term we intend to create a plugin system. We could solve exactly this problem using a plugin system that people opt-in to, which would execute some additional code. The code would be conditionally injected into bash (by javascript) using a variable.

Even then we could still keep the loop to iterate over PlayMode and EditMode.

Going with your current approach:

Right now because of the following code it would never combine the results in case I were to choose PlayMode only.

```ts
    const testPlatforms = (
      testMode === 'all' ? ['playmode', 'editmode', 'COMBINE_RESULTS'] : [testMode]
    ).join(';');
```

But does that mean the report is still generated? Or does the generation solely happen inside that combine results step?

A) If generation would work like this then I'd say lets go with this approach. 

B) If it does not work and it requires additional logic, perhaps we can inject the whole command as if it were a plugin from TS into the shell script using a variable already.

Our plugins system at its simplest would look roughly like this (consider psuedo) (I didn't get to implementing it yet)

```ts
// typescript:

  const runEditor = (args) => exec(`
   unity-editor \
     -batchmode \
    -logFile "$FULL_ARTIFACTS_PATH/$platform.log" \
    -projectPath "$UNITY_PROJECT_PATH" \
   ${args}`


  class Plugin { 
    static editorCall() {
      runEditor(`
        -coverageResultsPath "$FULL_COVERAGE_RESULTS_PATH" \
        -enableCodeCoverage \
        -debugCodeOptimization \
        -coverageOptions "$COVERAGE_OPTIONS" \
        // omitted CUSTOM_PARAMETERS, as i'ts only a combine step that's internal to the plugin itself
      `)
    }
  }

  // somewhere after docker.run()
 for(const plugin of pluginsRegistry) {
  plugin.postRuntimeCalls?.()
 }
```

and omit it from bash entirely (or even leave the bash script untouched entirely).

Now I don't expect you to implement all that, but I'd love your thoughts on this, as well as on what would be the best intermediate step in case (A) is not applicable.

nicholas-maltbie replied:
Actually option (A) is how it works. Right now, since we use the default `-coverageOptions` as provide it would work like this (when coverage options = `enableCyclomaticComplexity;generateHtmlReport;generateBadgeReport`)

When running just `playmode` or `editmode`, it will generate an html report and coverage results

When running `all` it will do the following steps
1. Generate a html report and coverage results for `playmode`.
2. Overwrite the previous html report for `playmode` and make a new coverage results for only `editmode`. 
3. The `COMBINE_RESULTS` run will overwrite the previous html report for `editmode` and create a new combined report from the results of `editmode` and `playmode` as well as a combined coverage file. 

The flags `generateHtmlReport` and `generateBadgeReport` tell the code coverage program to generate and overwrite the existing report with the results of the current test. since when we use `COMBINE_REUSLTS` we only have the `-quit` and `-testResults` flags, unity knows to combine all the current results in the output folder. You can see an example of this in the test runs. I'll include the code coverage reports in the `main.yml` example file for running in each type of mode actually (just upload the `CodeCoverage` output folder as well). 

Also, unity supports running more test profiles than just `playmode` or `editmode` if the user defines custom ones so hopefully this could allow for more user customization if the end user wants to in the future. 

I think adding a `Plugin` run in the future would probably be a better solution than the current `COMBINE_REUSLTS` but we can leave that to a future PR. Maybe instead of injecting the flags for what modes to run via environment variables, we could create an XML or variables file, pass it into the container, and change the bash script to iterate through each element in the XML file and read the customer arguments based on the values in the XML. would be easy to parse and run, easy to add more customization and debugging? 

I can open up an issue to start investigating how to structure those XML files and come up with a basic format for these runs. It could even be extended to the other plugins like the `unity-builder` if we would ever need to run multiple builds as part of one run.

webbertakken commented on src/model/output.test.ts:
```diff
@@ -6,4 +6,9 @@ describe('Output', () => {
   });
+  describe('setCoverageResultsPath', () => {
+    it('does not throw', async () => {
+      await expect(Output.setCoverageResultsPath('')).resolves.not.toThrow();
```
Lets test with an actual path and not an empty string.

nicholas-maltbie replied:
I'll add some more test, good point :)

nicholas-maltbie replied:
Added some tests 😄

webbertakken reviewed (changes requested):
Great initiative! Thanks for the PR.

It looks mostly good to me but I'd prefer to be very careful about extending the input parameters (as they're susceptible for breaking changes) and not add any more logic to the bash scripts.

nicholas-maltbie commented on action.yml:
```diff
@@ -20,6 +20,18 @@ inputs:
     required: false
     default: 'all'
     description: 'The type of tests to be run by the test runner.'
+  enableCodeCoverage:
+    required: false
+    default: 'false'
+    description: 'Should code coverage results be collected for the test run.'
```
Good point, I'll change it to just `coverage` as that makes a lot more sense.

nicholas-maltbie replied:
One reason to include this as a public api is that it requires the `-debugCodeOptimization` and `-enableCodeCoverage` flags. But now that I write this down, it would make more sense to always enable coverage by default and have a separate flag to turn off the `-debugCodeOptimization` or `-enableCodeCoverage`.  Since this is for running tests, most end users probably won't mind the `-debugCodeOptimization` but if someone wanted to run something like a performance test it would help. But I think you probably would need a full build for that anyway.

Given all that reasoning, let's leave it as enabled by default, I think I like that option better! It also makes some of the other changes much simpler.

nicholas-maltbie commented on .github/workflows/main.yml:
```diff
@@ -209,6 +209,62 @@ jobs:
+    strategy:
+      fail-fast: false
+      matrix:
+        enableCodeCoverage:
+          - true
+        unityVersion:
+          - 2019.2.11f1
+        projectPath:
+          - unity-project-with-correct-tests
+        coverageAssemblyFilters:
+          - '+MyScripts'
```
Yes, it is much easier to read this way! Should we keep the `projectPath` and `unityVersion` parameters since they are in the other workflow's matrices as well?

nicholas-maltbie commented on dist/steps/run_tests.sh:
```diff
@@ -36,6 +44,32 @@ case $TEST_MODE in
     ;;
 esac
 
+#
+# Build code coverage parameters
+#
+CODE_COVERAGE_PARAMETERS=""
+if [ "$ENABLE_CODE_COVERAGE" = "true" ]; then
+  # Configure code coverage options
+  COVERAGE_OPTIONS=""
+  ASSEMBLY_FILTER_OPTIONS=""
+
+  # Setup assembly filters if provided
+  if [ -n "$COVERAGE_ASSEMBLY_FILTERS" ]; then
+    ASSEMBLY_FILTER_OPTIONS=";assemblyFilters:$COVERAGE_ASSEMBLY_FILTERS"
+  fi
+
+  # Options to combine both playmode and editmode results
+  if [ "$EDIT_MODE" = "true" ] && [ "$PLAY_MODE" = "true" ]; then
+    COVERAGE_OPTIONS="enableCyclomaticComplexity$ASSEMBLY_FILTER_OPTIONS"
+  else
+    COVERAGE_OPTIONS="enableCyclomaticComplexity;generateHtmlReport;generateBadgeReport$ASSEMBLY_FILTER_OPTIONS"
+  fi
+
+  # Set parameters for code coverage
+  CODE_COVERAGE_PARAMETERS="-debugCodeOptimization -enableCodeCoverage -coverageOptions $COVERAGE_OPTIONS -coverageResultsPath $FULL_COVERAGE_RESULTS_PATH"
+fi
+echo "Using code coverage parameters $CODE_COVERAGE_PARAMETERS."
+
```
Ok, we will need to adjust the bash script somewhat to be able to combine the results from editmode and playmode to combine the results https://docs.unity3d.com/Packages/com.unity.testtools.codecoverage@1.1/manual/CoverageBatchmode.html#generate-combined-report-from-editmode-and-playmode-tests

However, given the comment from https://github.com/game-ci/unity-test-runner/pull/182#discussion_r852305527 I think we can adjust the bash script to be more modular.

nicholas-maltbie commented:
Hello @webbertakken , thanks for the review, I updated the PR based on your notes.

I left the comments about the default coverage output path, and bash script changes. I changed it to `coverage` but should the default be set to artifacts instead? The normal user probably wouldn't mind the extra data if it pertains to understanding the test results so that extra parameter can be removed if we don't need it. 

You also mentioned that we don't want to change the bash script. However, we need to run the unity editor with a few special flags to combine the results from an assembly run. I added a special keyword for `COMBINE_COVERAGE` as a platform to combine results of previous runs at the end. Here is the example of combining coverage from the unity docs:
```PowerShell
Unity.exe -projectPath <path-to-project> -batchmode -testPlatform editmode -runTests -debugCodeOptimization -enableCodeCoverage -coverageResultsPath <path-to-coverage-results>
-coverageOptions "generateAdditionalMetrics;assemblyFilters:+my.assembly.*"

Unity.exe -projectPath <path-to-project> -batchmode -testPlatform playmode -runTests -debugCodeOptimization -enableCodeCoverage -coverageResultsPath <path-to-coverage-results>
-coverageOptions "generateAdditionalMetrics;assemblyFilters:+my.assembly.*"

Unity.exe -projectPath <path-to-project> -batchmode -debugCodeOptimization -enableCodeCoverage -coverageResultsPath <path-to-coverage-results>
-coverageOptions "generateHtmlReport;generateBadgeReport;assemblyFilters:+my.assembly.*" -quit
```
As described in the comment, I changed the `run_tests.sh` script to use a loop instead of repeated blocks of if statements. These changes cover this special scenario as well as simply the code to run some or all tests. Although if we really don't want to edit the bash scripts, then I'm sure we could work out another solution.

timcassell commented on dist/steps/run_tests.sh:
```diff
@@ -65,118 +56,61 @@ echo ""
-  cat "$FULL_ARTIFACTS_PATH/editmode-results.xml"
-  cat "$FULL_ARTIFACTS_PATH/editmode-results.xml" | grep test-run | grep Passed
-fi
+  if [ $TEST_EXIT_CODE -gt 0 ]; then
```
While we're at it, this should be `-ne 0` since failed exit codes can be negative.

nicholas-maltbie replied:
I didn't know that I'll add the change!

webbertakken reviewed (approved):
After the last two comments are addressed I think this is good to go.

webbertakken commented:
> > As described in the comment, I changed the `run_tests.sh` script to use a loop instead of repeated blocks of if statements. These changes cover this special scenario as well as simply the code to run some or all tests. Although if we really don't want to edit the bash scripts, then I'm sure we could work out another solution.
> 
> I imagine that will also help in the future when Unity officially supports Docker and we can get Standalone tests working.

Ultimately we'd prefer to not inject any bash scripts into the image. If Unity has official support for docker our actions would still contain a lot of logic. We're looking to consolidate this logic (across all actions) into a CLI. The CLI would bootstrap a docker image (whether it's ours or our image is replaced by Unity's official docker image). The action would then just call the CLI.

On discord we have shared some diagrams that describe our general direction in the #cross-ci-bootstrapper channel.

nicholas-maltbie commented:
Hey @webbertakken and @timcassell Thanks for the reviews on the PR.

I added your suggested changes and have left a bit more discussion about future parameter changes for code coverage (https://github.com/game-ci/unity-test-runner/pull/182#discussion_r853689493) and how we could run this variable in the future via injecting a run settings xml file instead of passing environment variables (https://github.com/game-ci/unity-test-runner/pull/182#discussion_r853693518). Possible future ideas but I'll leave them for a future PR as it's better to keep code changes self-contained and smaller if possible. 

I made a test PR on my fork to verify the new changes work as expected. You can inspect some of the coverage results uploaded from the workflow runs (I have an example for `editmode`, `playmode`, and `all`).

If you have any more suggestions or comments let me know; otherwise, I think these changes should be ready!

If we merge these changes, I can also add some notes to the documentation project on how to add and configure the code coverage results for the github actions workflow as well, since the `coverageOptions` input is quite complex and probably needs some explaining.

webbertakken reviewed (approved):
Do we know if running tests with coverage takes much longer?

As I understand it the feature is now enabled by default, which is great as long as it doesn't slow down large test workflows too much.

If this is no concern, then this looks good to me.

nicholas-maltbie commented:
> Do we know if running tests with coverage takes much longer?
> 
> As I understand it the feature is now enabled by default, which is great as long as it doesn't slow down large test workflows too much.
> 
> If this is no concern, then this looks good to me.

Adding test coverage does not significantly slow down the tests. However, generating an HTML report for a large project can take quite a bit of time and generate a lot of files. We could leave the HTML report disabled by default and let the end user add the parameter in the `coverageOptions` if they want to include it. The main overhead is that we run the test in the editor instead of from a pre-compiled version of the code. According to the docs there is a small overhead, but I had a project with over 100 tests in the past and I never noticed a large difference between running with or without code coverage performance. 
> Note: Enabling Code Coverage adds some overhead to the editor and can affect the performance.

https://docs.unity3d.com/Packages/com.unity.testtools.codecoverage@1.2/manual/CoverageRecording.html?q=performance

Maybe we could make a follow up PR to add some flags and upload the coverage results in an XML format like is done for the test results. 
![image](https://user-images.githubusercontent.com/3240136/164344293-4b4b8f33-3a87-48c0-844a-8d2c99dfa5ff.png)

webbertakken commented:
If the overhead is small then I think we're good.

I like having this enabled by default so that new developers get this functionality out of the box.

If indeed there are projects that are negatively affected by this then we can still add a parameter called `coverage` which defaults to `true`, so that it can be disabled if needed. For now lets assume everyone profits from code coverage.

nikosatwork commented on action.yml:
```diff
@@ -20,6 +20,10 @@ inputs:
     description: 'The type of tests to be run by the test runner.'
+  coverageOptions:
+    required: false
+    default: 'enableCyclomaticComplexity;generateHtmlReport;generateBadgeReport'
```
The `enableCyclomaticComplexity` option has been replaced with the `generateAdditionalMetrics` option since version 0.3.0 of the Code Coverage package

Source: https://docs.unity3d.com/Packages/com.unity.testtools.codecoverage@1.2/changelog/CHANGELOG.html

webbertakken replied:
Updated in #183, release in [`v2.0.2`](https://github.com/game-ci/unity-test-runner/releases/tag/v2.0.2)

nicholas-maltbie commented on .github/workflows/main.yml:
```diff
@@ -105,6 +105,7 @@ jobs:
           projectPath: ${{ matrix.projectPath }}
           unityVersion: ${{ matrix.unityVersion }}
           testMode: all
+          coverageOptions: 'enableCyclomaticComplexity;generateHtmlReport;generateBadgeReport;assemblyFilters:+MyScripts'
```
Oh, good point. I'll make a small PR to update that parameter!

nicholas-maltbie commented:
> If the overhead is small then I think we're good.
> 
> I like having this enabled by default so that new developers get this functionality out of the box.
> 
> If indeed there are projects that are negatively affected by this then we can still add a parameter called `coverage` which defaults to `true`, so that it can be disabled if needed. For now lets assume everyone profits from code coverage.

I'm happy to add that parameter if needed in the future 😄

mikz commented:
I'm affected by this. Looks like Unity 2022.2.2f1 does not create the CodeCoverage folder. We don't have the code coverage package in the project.

```
2023-01-19T11:41:16.9010582Z Run succeeded, no failures occurred
2023-01-19T11:41:16.9010593Z 
2023-01-19T11:41:16.9010844Z ###########################
2023-01-19T11:41:16.9011123Z #    COMBINE_RESULTS Results    #
2023-01-19T11:41:16.9011377Z ###########################
2023-01-19T11:41:16.9011384Z 
2023-01-19T11:41:17.2783680Z chmod: cannot access '/github/workspace/CodeCoverage': No such file or directory
2023-01-19T11:41:17.2784978Z Changing to "/github/workspace/_activate-license" directory.
2023-01-19T11:41:17.2786410Z /github/workspace/_activate-license /github/workspace
2023-01-19T11:41:17.2787286Z /github/workspace
```

Sunbvert commented:
Looks like the example yml on the Test runner page is not working anymore for codecoverage. CodeCoverage folder is not generated.

webbertakken commented:
@Sunbvert Would you mind creating a new issue instead?

Sunbvert commented:
@webbertakken okay, will do it after confirming it.
